src/config.rs

```diff
@@ -85,23 +106,82 @@ impl LayeredConfigs {
 
     /// Creates new merged config.
     pub fn merge(&self) -> config::Config {
-        let config_sources = [
-            Some(&self.default),
-            Some(&self.env_base),
-            self.user.as_ref(),
-            self.repo.as_ref(),
-            Some(&self.env_overrides),
-            self.arg_overrides.as_ref(),
-        ];
-        config_sources
+        self.sources()
             .into_iter()
-            .flatten()
+            .map(|(_, config)| config)
             .fold(config::Config::builder(), |builder, source| {
                 builder.add_source(source.clone())
             })
             .build()
             .expect("loaded configs should be merged without error")
     }
+
+    fn sources(&self) -> Vec<(ConfigSource, &config::Config)> {
+        let config_sources = [
+            (ConfigSource::Default, Some(&self.default)),
+            (ConfigSource::Env, Some(&self.env_base)),
+            (ConfigSource::User, self.user.as_ref()),
+            (ConfigSource::Repo, self.repo.as_ref()),
+            (ConfigSource::Env, Some(&self.env_overrides)),
+            (ConfigSource::CommandArg, self.arg_overrides.as_ref()),
+        ];
+        config_sources
+            .into_iter()
+            .filter_map(|(source, config)| config.map(|c| (source, c)))
+            .collect_vec()
+    }
+
+    pub fn resolved_config_values(
+        &self,
+        filter_prefix: &[&str],
+    ) -> Result<Vec<AnnotatedValue>, ConfigError> {
+        // Collect annotated values from each config.
+        let mut config_vals = vec![];
+
+        for (source, config) in self.sources() {
+            let config_table = config.collect()?.into();
+            let mut config_stack: Vec<(Vec<&str>, &config::Value)> = vec![(vec![], &config_table)];
+            while let Some((path, value)) = config_stack.pop() {
+                // Skip if path doesn't match filter_prefix.
+                if filter_prefix.iter().zip(path.iter()).any(|(a, b)| *a != *b) {
+                    continue;
+                }
+                match &value.kind {
+                    config::ValueKind::Table(table) => {
+                        // TODO: Remove sorting when config crate maintains deterministic ordering.
+                        for (k, v) in table.iter().sorted_by_key(|(k, _)| *k).rev() {
+                            let mut key_path = path.to_vec();
+                            key_path.push(k);
+                            config_stack.push((key_path, v));
+                        }
+                    }
+                    _ => {
+                        config_vals.push(AnnotatedValue {
+                            path: path.iter().map(|s| s.to_string()).collect_vec(),
```

Comment:
I don't love cloning to `String`s here, but had trouble figuring out lifetimes to use `&str`. Any suggestions? Or is it more trouble than it's worth?

Reply:
Since `config_table` is instantiated locally, it wouldn't be possible to return `&str` that references the `config_table`. Maybe you can eliminate some of `.to_string()` by consuming the source `ValueKind::Table()`, but that wouldn't worth the effort.

Reply:
The catch is I'd have to explicitly deref to use `to_owned()` since it's a `&&str` where `s.to_string()` handles all the derefs implicitly. Is there a way to de-uglyify `path.iter().map(|s| (*s).to_owned()).collect_vec()` (without making throwaway copies of the whole vector)?

Reply:
`|&s| s.to_owned()` is one option. `.into_iter()` should also work, but clippy might complain.

fwiw, I agree `.to_owned()` is preferred so long as the type can be inferred.